### PR TITLE
Fix NewFileLspSuite on Windows

### PR DESCRIPTION
**Pull request by tgodzik**
Getting weird access denied exceptions. Might be a race condition between folder being opened and used.
```
INFO  Executing command [new-scala-file]
	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
[error] ==> X tests.NewFilesLspSuite.new-trait-new-dir  2.929s java.nio.file.AccessDeniedException: C:\Users\RUNNER~1\AppData\Local\Temp\metals14694223127987732815Baz.scala -> D:\a\metals\metals\tests\unit\target\e2e\new-files\new-trait-new-dir\a\src\main\scala\bar\Baz.scala
	at sun.nio.fs.WindowsFileSystemProvider.newByteChannel(WindowsFileSystemProvider.java:235)
[info] tests.NewFilesLspSuite.new-package-object started
	at java.nio.file.Files.newByteChannel(Files.java:370)
	at java.nio.file.Files.newByteChannel(Files.java:421)
	at java.nio.file.Files.readAllBytes(Files.java:3205)
	at scala.meta.internal.io.PlatformFileIO$.slurp(PlatformFileIO.scala:45)
	at scala.meta.internal.io.FileIO$.slurp(FileIO.scala:24)
	at scala.meta.internal.mtags.MtagsEnrichments$XtensionAbsolutePathMetals.toInput(MtagsEnrichments.scala:168)
	at scala.meta.internal.metals.Docstrings.expireSymbolDefinition(Docstrings.scala:58)
	at scala.meta.internal.metals.MetalsLanguageServer.indexSourceFile(MetalsLanguageServer.scala:1625)
	at scala.meta.internal.metals.MetalsLanguageServer.$anonfun$reindexWorkspaceSources$2(MetalsLanguageServer.scala:1592)
	at scala.meta.internal.metals.MetalsLanguageServer.$anonfun$reindexWorkspaceSources$2$adapted(MetalsLanguageServer.scala:1589)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.meta.internal.metals.MetalsLanguageServer.reindexWorkspaceSources(MetalsLanguageServer.scala:1589)
	at scala.meta.internal.metals.MetalsLanguageServer.$anonfun$onChange$2(MetalsLanguageServer.scala:962)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
```

### CODE REVIEW

**Comment by kpbochenek** [comment]
Also error from latest build:
```
ERROR D:\a\metals\metals\tests\unit\target\e2e\new-files\new-trait-new-dir\a\src\main\scala\bar\Baz.scala
INFO  time: indexed workspace in 1.38s
INFO  tests.TestingServer#executeCommand
java.nio.file.FileSystemException: D:\a\metals\metals\tests\unit\target\e2e\new-files\new-trait-new-dir\a\src\main\scala\bar\Baz.scala: The process cannot access the file because it is being used by another process.
INFO  Executing command [new-scala-file]

[info]   + new-trait-new-dir 3.601s
	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
[info] tests.NewFilesLspSuite.new-package-object started
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at sun.nio.fs.WindowsFileSystemProvider.newByteChannel(WindowsFileSystemProvider.java:235)
```

Isn't this actually a bug not a flakiness?

**Comment by tgodzik** [comment]
@kpbochenek You might be right, I think there is a possibility there was a race condition, where we try to create a file in two ways. We should be fine to just use `writeText`, which already creates everything. Let's see if it helps.

**Comment by tgodzik** [comment]
First run clean, reruning...

**Comment by tgodzik** [comment]
Second one clear, 3 time is the charm.

**Comment by tgodzik** [comment]
It looks like the issue is fixed, but not 100% sure. This change is still valid even if it doesn't fix anything.

**Comment by kpbochenek** [comment]
Looking at implementation of `writeText` it looks like it creates all necessary parent directories, creates tmp file, puts code there and then performs `move` where a file should actually be.

With that in mind I think `createFile` call is not necessary here.
Also exception showing here might be caused by metals/bloop locking a freshly created file, then we create temp file and put content there, and when we perform move we fail because a file is used by bloop/metals?

**Comment by kpbochenek** [comment]
Also I just realized we don't check anywhere if a file already exists if we want to create a new class :fearful: 

Which means you can easily erase your whole file by creating a class with the same name :fearful: 

Do you agree `new-scala-file` should fail if file already exists? (and not erase currently existing file? :fearful: )

**Comment by alekseiAlefirov** [comment]
> Do you agree new-scala-file should fail if file already exists? (and not erase currently existing file? fearful )

yes, I think, it should fail (and mention it to the user) and not rewrite file.

File: tests/unit/src/test/scala/tests/NewFilesLspSuite.scala [hunk 1]
```diff
@@ -102,7 +103,7 @@ class NewFilesLspSuite extends BaseLspSuite("new-files") {
 
   private def indent = "  "
 
-  private def check(testName: String)(
+  private def check(testName: TestOptions)(
```
Comment (alekseiAlefirov) [inline tests/unit/src/test/scala/tests/NewFilesLspSuite.scala:106]: just out of curiosity:
1) why?
2) Is there implicit conversion happening? I see, there is `TestOptionsConversions` at munit, but it does not look imported and available...

**Comment by alekseiAlefirov** [comment]
@tgodzik so you saying, that two `Files.createFile` sequential calls eventually get asynchronous?
Sounds like a source of flakiness :neutral_face:

File: tests/unit/src/test/scala/tests/NewFilesLspSuite.scala (hunk 1 above)
Comment (tgodzik) [reply tests/unit/src/test/scala/tests/NewFilesLspSuite.scala:106]: I can revert, but this enables adding tags and ignoring.

**Comment by tgodzik** [comment]
> Looking at implementation of writeText it looks like it creates all necessary parent directories, creates tmp file, puts code there and then performs move where a file should actually be.
> With that in mind I think createFile call is not necessary here.
> Also exception showing here might be caused by metals/bloop locking a freshly created file, then we create temp file and put content there, and when we perform move we fail because a file is used by bloop/metals?

I think that was an issue, trying to create a file 2 different ways at the same time. Windows might have locked some directories and that's why the exception. I change to just using `writeText`, which should be enough.

> Also I just realized we don't check anywhere if a file already exists if we want to create a new class 
> Which means you can easily erase your whole file by creating a class with the same name fearful
> Do you agree new-scala-file should fail if file already exists? (and not erase currently existing file? fearful )

I will add a check, good idea!

> @tgodzik so you saying, that two Files.createFile sequential calls eventually get asynchronous?
> Sounds like a source of flakiness 

The lock might not be released in time, that might have sometimes cuased flaky issues.

File: tests/unit/src/test/scala/tests/NewFilesLspSuite.scala (hunk 1 above)
Comment (ckipp01) [reply tests/unit/src/test/scala/tests/NewFilesLspSuite.scala:106]: We also do this in other places, so we probably don't have to revert.

**Comment by tgodzik** [comment]
I added a check to see if a file exists and an accompanying test case.

File: metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala [hunk 2]
```diff
@@ -164,15 +164,22 @@ class NewFilesProvider(
...
-    val result = Future {
-      path.touch()
-      path
+  private def createFileAndWriteText(
+      path: AbsolutePath,
+      template: NewFileTemplate
+  ): Future[(AbsolutePath, Range)] = {
+    val result = if (path.exists) {
+      Future.failed(
+        new IllegalArgumentException(s"The file $path already exists.")
```
Comment (alekseiAlefirov) [inline metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala:176]: maybe `FileAlreadyExistsException`?

File: tests/unit/src/test/scala/tests/NewFilesLspSuite.scala [hunk 3]
```diff
@@ -100,15 +101,36 @@ class NewFilesLspSuite extends BaseLspSuite("new-files") {
...
   )
 
+  val filePath = "a/src/main/scala/foo/Other.scala"
+  val existingFileContents =
+    """|package foo
+       |
+       |class Other {
+       | val withContent = true
+       |}
+       |""".stripMargin
```
Comment (alekseiAlefirov) [inline tests/unit/src/test/scala/tests/NewFilesLspSuite.scala:111]: these two declarations look a bit out here, among `check`s. I would put it somewhere down with appropriate names (fixture style), or remove a new line between them and the check below, to express they are related.

Comment (tgodzik) [reply tests/unit/src/test/scala/tests/NewFilesLspSuite.scala]: I moved them into the check, I think it's fine to have it duplicated there and looks more like the other test cases.

File: metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala (hunk 2 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala]: Good idea! Changed

**Pull request merged**